lib/Push.php

```diff
@@ -0,0 +1,164 @@
+use OCP\Notification\INotification;
+
+class Push {
+	/** @var IDBConnection */
+	protected $connection;
+	/** @var IManager */
+	protected $manager;
+	/** @var IConfig */
+	protected $config;
+	/** @var IProvider */
+	protected $tokenProvider;
+	/** @var Manager */
+	private $keyManager;
+	/** @var IUserManager */
+	private $userManager;
+	/** @var IClientService */
+	protected $clientService;
+	/** @var ILogger */
+	protected $log;
+
+	public function __construct(IDBConnection $connection, IManager $manager, IConfig $config, IProvider $tokenProvider, Manager $keyManager, IUserManager $userManager, IClientService $clientService, ILogger $log) {
+		$this->connection = $connection;
+		$this->manager = $manager;
+		$this->config = $config;
+		$this->tokenProvider = $tokenProvider;
+		$this->keyManager = $keyManager;
+		$this->userManager = $userManager;
+		$this->clientService = $clientService;
+		$this->log = $log;
+	}
+
+	/**
+	 * @param INotification $notification
+	 */
+	public function pushToDevice(INotification $notification) {
+		$devices = $this->getDevicesForUser($notification->getUser());
+		$user = $this->userManager->get($notification->getUser());
+
+		if (empty($devices) || !($user instanceof IUser)) {
+			return;
+		}
+
+		$language = $this->config->getUserValue($notification->getUser(), 'core', 'lang', 'en');
+		try {
+			$notification = $this->manager->prepare($notification, $language);
+		} catch (\InvalidArgumentException $e) {
+			return;
+		}
+
+		$userKey = $this->keyManager->getKey($user);
+
+		$collection = [];
+		foreach ($devices as $device) {
+			try {
+				$collection[] = $this->encryptAndSign($userKey, $device, $notification);
+			} catch (InvalidTokenException $e) {
+				// Token does not exist anymore, should drop the push device entry
+				// FIXME delete push token
+			} catch (\InvalidArgumentException $e) {
+				// Token does not exist anymore, should drop the push device entry
+				// FIXME delete push token
+			}
+		}
+
+		$payload = json_encode($collection);
+
+		$this->log->alert($payload); // TODO TEMP
+
+		$client = $this->clientService->newClient();
+		try {
+			$response = $client->post('http://127.0.0.1:3306', [
+				'body' => $payload,
+			]);
+		} catch (\Exception $e) {
+			$this->log->logException($e, [
+				'app' => 'notifications',
+			]);
+			return;
+		}
+	}
+
+	/**
+	 * @param Key $userKey
+	 * @param array $device
+	 * @param INotification $notification
+	 * @return array
+	 * @throws InvalidTokenException
+	 * @throws \InvalidArgumentException
+	 */
+	protected function encryptAndSign(Key $userKey, array $device, INotification $notification) {
+		// Check if the token is still valid...
+		$this->tokenProvider->getTokenById($device['token']);
+
+		$data = [
+			'app' => $notification->getApp(),
+			'subject' => $notification->getParsedSubject(),
+		];
+
+		if (!openssl_private_encrypt(json_encode($data), $encryptedSubject, $device['devicepublickey'], OPENSSL_PKCS1_PADDING)) {
```

Comment:
@LukasReschke are this call and the one to openssl_sign below correct? Would like to know, before I start failing on golang again....